c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -6647,66 +6668,310 @@ namespace URLSearchParams {
+    UniqueChars backend_chars = encode(cx, backend, &backend_len);
+    if (!backend_chars)
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+
+    PendingRequestHandle request_handle = { INVALID_HANDLE };
+    RootedObject response_promise(cx, JS::NewPromiseObject(cx, nullptr));
+    if (!response_promise)
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+
+    // TODO: ensure that we properly handle body handles forwarded from other Request/Response
+    // objects.
+    int result = xqd_req_send_async(Request::request_handle(request),
+                                    RequestOrResponse::body_handle(request),
+                                    backend_chars.get(), backend_len, &request_handle);
+    if (!HANDLE_RESULT(cx, result))
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+
+    if (!pending_requests->append(request))
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+
+    JS::SetReservedSlot(request, Request::Slots::PendingRequest, JS::Int32Value(request_handle.handle));
+    JS::SetReservedSlot(request, Request::Slots::ResponsePromise, JS::ObjectValue(*response_promise));
+
+    args.rval().setObject(*response_promise);
+    return true;
+  }
+
+  /**
+   * The `queueMicrotask` global function
+   * https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#microtask-queuing
+   */
+  bool queueMicrotask(JSContext* cx, unsigned argc, Value* vp) {
+    CallArgs args = CallArgsFromVp(argc, vp);
+    if (!args.requireAtLeast(cx, "queueMicrotask", 1)) {
+      return false;
+    }
+
+    if (!args[0].isObject() || !JS::IsCallable(&args[0].toObject())) {
+      JS_ReportErrorLatin1(cx, "queueMicrotask: Argument 1 is not a function");
+      return false;
+    }
+
+    RootedObject callback(cx, &args[0].toObject());
+
+    RootedObject promise(cx, JS::CallOriginalPromiseResolve(cx, JS::UndefinedHandleValue));
+    if (!promise) {
+      return false;
+    }
+
+    if (!JS::AddPromiseReactions(cx, promise, callback, nullptr)) {
+      return false;
+    }
+
+    args.rval().setUndefined();
+    return true;
+  }
+
+  bool self_get(JSContext* cx, unsigned argc, Value* vp) {
+    CallArgs args = CallArgsFromVp(argc, vp);
+    args.rval().setObject(*JS::CurrentGlobalOrNull(cx));
+    return true;
+  }
+
+  bool self_set(JSContext* cx, unsigned argc, Value* vp) {
+    CallArgs args = CallArgsFromVp(argc, vp);
+    if (!args.requireAtLeast(cx, "globalThis.self setter", 1)) {
+      return false;
+    }
+
+    RootedObject global(cx, JS::CurrentGlobalOrNull(cx));
+    if (args.thisv() != ObjectValue(*global)) {
+      JS_ReportErrorLatin1(cx, "globalThis.self setter can only be called on the global object");
+      return false;
+    }
+
+    if (!JS_DefineProperty(cx, global, "self", args[0], JSPROP_ENUMERATE)) {
+      return false;
+    }
+
+    args.rval().setUndefined();
+    return true;
+  }
+
+  #define SCTAG_DOM_URLSEARCHPARAMS 0xffff8014 // Same tag as used by Gecko.
```

Comment:
A slightly longer comment here might help understanding -- something like "Magic number used in structured cloning as a tag to identify a URLSearchParam"?

Also, regarding use of the same tag as in Gecko -- could you add a comment why? Is it just a convenience to use the same tag-value assignments (and we could choose any value but we arbitrarily choose this one), or is there some expectation that the serialized form could be exposed someday and the tag/format matters?

Reply:
Both very good points! :) There is absolutely no technical reason to reuse Gecko's value here—I just didn't want to think about it 😬 But cargo-culting without explanation is annoying, so I changed this to the first available tag, conveniently identified by `JS_SCTAG_USER_MIN`, and added the comment you suggested.